src/main/java/seedu/address/logic/commands/UndoCommand.java

```diff
@@ -31,17 +37,78 @@ public class UndoCommand extends Command {
     @Override
     public CommandResult execute(Model model) throws CommandException {
         requireNonNull(model);
-        Person deletedPerson = model.getDeletedPerson();
 
-        //catch multiple undo statements done in a row.
-        if (model.hasPerson(deletedPerson)) {
-            throw new CommandException(MESSAGE_UNDO_FAILURE);
+        if (model.getPreviousUndoableCommandsSize() == 0) {
+            throw new CommandException(MESSAGE_FAILURE_UNDO_WITH_NO_UNDOABLE_COMMANDS);
         }
 
-        model.addPerson(deletedPerson);
-        return new CommandResult(String.format(MESSAGE_UNDO_SUCCESS, Messages.format(deletedPerson)));
+        switch(model.getPreviousUndoableCommand()) {
+        case "delete":
+            return executeUndoDelete(model);
+        case "clear":
+            return executeUndoClear(model);
+        case "add":
+        case "clone":
+            return executeUndoAdd(model);
+        case "edit":
+            return executeUndoEdit(model);
+        default:
+            return null;
```

Comment:
Does returning null result in any errors to the function it pass the value to? I had problem once with appointment date for returning null values if no dates are indicated

Reply:
I've amended this by throwing an AssertionError for the default case in my latest commit. :D